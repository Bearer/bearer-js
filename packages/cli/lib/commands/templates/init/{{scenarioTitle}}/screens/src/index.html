<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
    <title>{{scenarioTitle}} scenario</title>
    <script src="/build/{{componentTagName}}.js"></script>
    <style>
      body {
        padding: 50px;
      }
      hr {
        margin: 20px 0;
      }
    </style>
  </head>
  <body>
    <h1>Setup component</h1>
    <{{componentTagName}}-setup></{{componentTagName}}-setup>
    <hr />
    <h1>Action component</h1>
    <{{componentTagName}}></{{componentTagName}}>
    <hr />
    <h1>Display component</h1>
    <{{componentTagName}}-display></{{componentTagName}}-display>
  </body>
</html>
