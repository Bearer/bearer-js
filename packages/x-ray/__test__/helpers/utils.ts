import EventEmitter from 'events'

export class TestEmitter extends EventEmitter {
  method: string = 'GET'
  url: string = '/'
  connection: any = {}
  req: any = {}
  statusCode: number = 200
  statusMessage: string = 'KO'

  resume() {
    this.emit('resume')
  }
}

function buildFakeRequest() {
  const request = new TestEmitter()
  request.method = 'GET'
  request.url = '/'
  request.connection = { remoteAddress: 'myhost' }
  return request
}

function buildFakeResponse() {
  const response = new TestEmitter()
  response.statusCode = 200
  response.statusMessage = 'OK'

  response.resume = () => {
    response.emit('end')
  }
  return response
}

export const httpClient = {
  request: (options: any, callback: Function) => {
    const fakeRequest = buildFakeRequest()
    const fakeReponse = buildFakeResponse()
    fakeReponse.req = fakeRequest
    callback(fakeReponse)
    return fakeRequest
  },
  get: (options: any, callback: Function) => {
    const fakeRequest = buildFakeRequest()
    const fakeReponse = buildFakeResponse()
    fakeReponse.req = fakeRequest
    callback(fakeReponse)
    return fakeRequest
  }
}

export const expectedResponse = {
  message: {
    clientId: '132464737464748494404949984847474848',
    integrationUuid: 'scenarioUuid',
    intentName: 'MyHandler',
    method: 'GET',
    path: 'www.google.com',
    pathname: '/',
    responseStatus: 200,
    responseStatusMesage: 'OK',
    stage: 'test'
  },
  timestamp: expect.any(Number)
}
