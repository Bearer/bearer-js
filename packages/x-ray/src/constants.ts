export const { _HANDLER, STAGE, BEARER_XRAY_LOG_GROUP, AWS_LAMBDA_LOG_STREAM_NAME } = process.env
