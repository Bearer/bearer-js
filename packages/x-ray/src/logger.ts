import debug from '@bearer/logger'

export default debug('functions:xray')
