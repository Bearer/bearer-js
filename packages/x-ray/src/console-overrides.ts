import { STAGE, _HANDLER } from './constants'
import uuid = require('uuid')
import console from 'console'

import log from './logger'
import stream from 'stream'

const out = new stream.Writable()

out._write = (chunk, _encoding, next) => {
  userLogger('%j', buildLog(chunk.toString()))
  next()
}

const userLogger = log.extend('user')

const myConsole = new console.Console(out)

export const overrideConsole = (module: typeof console) => {
  const _log = module.log

  module.log = (...args: any) => {
    myConsole.log(args)

    return _log.apply(module, args)
  }
}

const buildLog = (args: any) => {
  return {
    buid: process.env.scenarioUuid,
    payloadType: 'DEBUG',
    service: 'function',
    payload: {
      log: args,
      intentName: _HANDLER,
      uuid: uuid()
    },
    stage: STAGE,
    tid: process.env._X_AMZN_TRACE_ID
  }
}
