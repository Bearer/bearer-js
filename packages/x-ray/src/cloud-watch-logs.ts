import CloudWatchLogs from 'aws-sdk/clients/cloudwatchlogs'
import uuid from 'uuid/v1'
import { AWS_LAMBDA_LOG_STREAM_NAME, BEARER_XRAY_LOG_GROUP } from './constants'
import logger from './logger'

const cloudWatchLogsClientInstance = new CloudWatchLogs()

export const sendToCloudwatchGroup = async (payload: any, cloudWatchLogsClient = cloudWatchLogsClientInstance) => {
  try {
    const streamName = AWS_LAMBDA_LOG_STREAM_NAME!.concat('-').concat(uuid())
    const event = {
      logGroupName: BEARER_XRAY_LOG_GROUP,
      logStreamName: streamName,
      logEvents: [
        {
          timestamp: payload.timestamp,
          message: JSON.stringify(payload.message)
        }
      ]
    } as CloudWatchLogs.Types.PutLogEventsRequest
    await createLogStream(cloudWatchLogsClient, streamName)
    await cloudWatchLogsClient.putLogEvents(event).promise()
  } catch (error) {
    console.log(`error ${error.message}`)
  }
}

const createLogStream = async (client: CloudWatchLogs, streamName: string) => {
  try {
    return await client
      .createLogStream({
        logGroupName: BEARER_XRAY_LOG_GROUP!,
        logStreamName: streamName
      })
      .promise()
  } catch (error) {
    logger('%j', error)
    return undefined
  }
}
