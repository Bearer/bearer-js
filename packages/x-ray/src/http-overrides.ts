// Mostly inspired from: https://github.com/meetearnest/global-request-logger/blob/master/index.js

import url from 'url'
import http from 'http'
import https from 'https'

import logger from './logger'
import { _HANDLER } from './constants'

const EXCLUDED_API = new Set(['logs.eu-west-3.amazonaws.com', 'logs.eu-west-1.amazonaws.com'])

export const overrideRequestMethod = (module: typeof http | typeof https) => {
  // override http.request method
  const _request = module.request

  function request(...args: any): http.ClientRequest {
    // @ts-ignore
    const req = _request(...arguments)
    const info = {
      request: {
        method: req.method || 'get',
        headers: { ...req.getHeaders() }
      } as { method: string; headers: any; error?: any; body: any } & url.UrlWithStringQuery,
      response: {} as {
        body: any
        statusCode: number
        headers: any
        error?: any
      }
    }
    const requestUrl = args[0]

    if (requestUrl === 'string') {
      const { port, path, host, protocol, auth, hostname, hash, search, query, pathname, href } = url.parse(requestUrl)
      info.request = {
        ...info.request,
        port,
        path,
        host,
        protocol,
        auth,
        hostname,
        hash,
        search,
        query,
        pathname,
        href
      }
    }
    const requestData: string[] = []
    const originalWrite = req.write

    req.write = function() {
      logBodyChunk(requestData, arguments[0])
      originalWrite.apply(req, arguments)
    }

    req.on('error', (error: any) => {
      info.request.error = error
    })

    req.on('response', (res: http.ServerResponse) => {
      info.request.body = requestData.join('')
      info.response.statusCode = res.statusCode
      info.response.headers = (res as any).headers

      const responseData: string[] = []
      res.on('data', (data: string) => {
        logBodyChunk(responseData, data)
      })

      res.on('end', () => {
        if (!EXCLUDED_API.has(info.request.host!)) {
          info.response.body = responseData.join('')
          logger.extend('externalCall')('%j', {
            buid: process.env.scenarioUuid,
            payloadType: 'FUNCTION',
            payload: info,
            service: 'function',
            tid: process.env._X_AMZN_TRACE_ID
          })
        }
      })
      res.on('error', (error: any) => {
        info.response.error = error
      })
    })

    return req
  }

  function get(...args: any): http.ClientRequest {
    // @ts-ignore
    const req = request(...arguments)
    req.end()
    return req
  }

  module.request = request
  module.get = get
}

const maxBodyLength = 2024 * 1000 * 3

function logBodyChunk(array: any[], chunk: string) {
  if (chunk) {
    let toAdd = chunk
    const newLength = array.length + chunk.length
    if (newLength > maxBodyLength) {
      toAdd = chunk.slice(0, maxBodyLength - newLength)
    }
    array.push(toAdd)
  }
}
