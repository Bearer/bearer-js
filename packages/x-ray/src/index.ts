import logger from './logger'
import { overrideRequestMethod } from './http-overrides'

let check: number

export const captureHttps = () => {
  const httpModule = require('http')
  const httpsModule = require('https')
  if (httpModule._bearerLoading === check && check) {
    logger('%j', { message: 'http module has already been loaded', application: 'x-ray' })
    return
  }
  httpModule._bearerLoading = check = Math.random()
  logger('%j', { message: 'Overriding request and get methods', application: 'x-ray' })
  overrideRequestMethod(httpModule)
  overrideRequestMethod(httpsModule)
}

export const setupFunctionIdentifiers = function(event: any) {
  const context = event.context || {}
  const { clientId, integrationUuid } = context
  logger('%j', { message: `ÃŒnject ${clientId} and ${integrationUuid}`, application: 'x-ray' })
  process.env.clientId = clientId
  process.env.scenarioUuid = integrationUuid
}

export default captureHttps
