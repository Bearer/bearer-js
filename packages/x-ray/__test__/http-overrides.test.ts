import http from 'http'
import { overrideGetMethod, overrideRequestMethod } from '../src/http-overrides'
import { httpClient } from './helpers/utils'
import logger from '../src/logger'

jest.mock('../src/logger')
jest.mock('../src/constants')
process.env.clientId = '132464737464748494404949984847474848'
process.env.scenarioUuid = 'scenarioUuid'

describe('override http', () => {
  it('overrides request method', () => {
    // @ts-ignore
    expect(httpClient._request).toBeUndefined()
    // @ts-ignore
    overrideRequestMethod(httpClient)
    // @ts-ignore
    expect(httpClient._request).toBeDefined()
  })

  it('traces request call with the right payload', async () => {
    // @ts-ignore
    expect(httpClient._request).toBeDefined()
    await new Promise((res, _rej) => {
      httpClient.request('http://www.google.com/', (data: http.IncomingMessage) => {
        res(data)
        data.resume()
      })
    })
  })

  it('overrides the get method', () => {
    // @ts-ignore
    expect(httpClient._get).toBeUndefined()
    overrideGetMethod(httpClient)
    // @ts-ignore
    expect(httpClient._get.name).toEqual('get')
    // @ts-ignore
    expect(httpClient._get).toBeDefined()
  })

  it('traces get call with the right payload', async () => {
    // @ts-ignore
    expect(httpClient.get).toBeDefined()
    await new Promise((res, _rej) => {
      httpClient.get('http://www.google.com/', (data: http.IncomingMessage) => {
        res(data)
        data.resume()
      })
    })
  })
})

describe('log filtering', () => {
  const spy = jest.spyOn(logger, 'extend')
  beforeEach(() => {
    spy.mockImplementationOnce(jest.fn())
    spy.mockReset()
  })

  it('is calling logger for non aws calls', async () => {
    await new Promise((res, _rej) => {
      httpClient.request('http://www.google.com/', (data: http.IncomingMessage) => {
        res(data)
        data.resume()
      })
    })

    expect(spy).toHaveBeenCalledWith('externalCall')
  })

  it('is not calling logger for eu-west-3 aws logs calls', async () => {
    await new Promise((res, _rej) => {
      httpClient.request('http://logs.eu-west-3.amazonaws.com/', (data: http.IncomingMessage) => {
        res(data)
        data.resume()
      })
    })

    expect(spy).not.toHaveBeenCalled()
  })

  it('is not calling logger for eu-west-1 aws logs calls', async () => {
    await new Promise((res, _rej) => {
      httpClient.request('http://logs.eu-west-1.amazonaws.com/', (data: http.IncomingMessage) => {
        res(data)
        data.resume()
      })
    })

    expect(spy).not.toHaveBeenCalled()
  })
})
