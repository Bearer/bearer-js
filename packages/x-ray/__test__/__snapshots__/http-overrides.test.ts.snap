// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overrideRequestMethod https get logs request 1`] = `
Array [
  "%j",
  Object {
    "message": Object {
      "clientId": "a-client-id",
      "integrationUuid": "custom0buid",
      "intentName": undefined,
      "method": "get",
      "path": "bearer.sh",
      "pathname": "/path",
      "stage": undefined,
      "tid": "aws_trace",
      "type": "externalCall",
    },
    "timestamp": 5,
  },
]
`;

exports[`overrideRequestMethod https get logs request 2`] = `
Array [
  "%j",
  Object {
    "buid": "custom0buid",
    "payload": Object {
      "duration": 10,
      "request": Object {
        "auth": null,
        "body": "",
        "hash": null,
        "headers": Object {},
        "host": "bearer.sh",
        "hostname": "bearer.sh",
        "href": "https://bearer.sh/path?ok=false",
        "method": "get",
        "path": "/path?ok=false",
        "pathname": "/path",
        "port": null,
        "protocol": "https:",
        "query": Object {
          "ok": "false",
        },
        "search": "?ok=false",
        "url": "https://bearer.sh/path?ok=false",
      },
      "response": Object {
        "body": "{\\"great\\":\\"sponge bob is the king\\"}",
        "headers": Object {
          "content-type": "application/json",
        },
        "status": 201,
      },
    },
    "payloadType": "EXTERNAL_CALL",
    "service": "function",
    "tid": "aws_trace",
  },
]
`;

exports[`overrideRequestMethod https request logs request 1`] = `
Array [
  "%j",
  Object {
    "message": Object {
      "clientId": "a-client-id",
      "integrationUuid": "custom0buid",
      "intentName": undefined,
      "method": "get",
      "path": "bearer.sh",
      "pathname": undefined,
      "stage": undefined,
      "tid": "aws_trace",
      "type": "externalCall",
    },
    "timestamp": 5,
  },
]
`;

exports[`overrideRequestMethod https request logs request 2`] = `
Array [
  "%j",
  Object {
    "buid": "custom0buid",
    "payload": Object {
      "duration": 15,
      "request": Object {
        "auth": undefined,
        "body": "{\\"todo\\":\\"Buy the milk\\"}",
        "hash": undefined,
        "headers": Object {
          "useragent": "Bearer",
        },
        "host": "bearer.sh:443",
        "hostname": "bearer.sh",
        "href": undefined,
        "method": "get",
        "path": "/postPath?postQuery=sponge",
        "pathname": undefined,
        "port": 443,
        "protocol": undefined,
        "query": undefined,
        "search": undefined,
        "url": "undefined//bearer.sh:443/postPath?postQuery=sponge",
      },
      "response": Object {
        "body": "{\\"great\\":\\"sponge bob is the king of posts\\"}",
        "headers": Object {
          "content-type": "application/json",
        },
        "status": 204,
      },
    },
    "payloadType": "EXTERNAL_CALL",
    "service": "function",
    "tid": "aws_trace",
  },
]
`;
